### PR TITLE
LSTM ONNX Layout Attribute Support 

### PR DESCRIPTION
**Pull request by Abdurrahheem**
### Explanation
This PR contains necessary changes to support `layout` attribute. This attributes is present in [ONNX](https://github.com/onnx/onnx/blob/main/docs/Operators.md#lstm) and [Torch](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#lstm) (in touch it is name as `batch_first=True`) libraries. When `layout = 1` input to LSTM layer is expected to have batch dimension first -> `[batch_size, sequence_length, features]` vs `layout = 0` - default `[sequence_length, batch_size, features]`

### Test Data
Test data and data generator for PR located here [#1063](https://github.com/opencv/opencv_extra/pull/1063)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@fengyuentau @dkurt Could you look at it?

**Review by fengyuentau — APPROVED** [review]
👍

File: modules/dnn/src/layers/recurrent_layers.cpp [hunk 1]
```diff
@@ -119,6 +119,8 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
     bool useCellClip, usePeephole;
     bool reverse;   // If true, go in negative direction along the time axis
     bool bidirectional;  // If true, produces both forward and reversed directions along time axis
+    bool layout;  // If true, uses batch_size x seq_length x num_hidden for input and output, else
```
Comment (dkurt) [inline modules/dnn/src/layers/recurrent_layers.cpp:122]: I suggest to create a enum. `layout` in ONNX has `int` type.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -1753,8 +1753,16 @@ void ONNXImporter::parseLSTM(LayerParams& layerParams, const opencv_onnx::NodePr
     CV_Assert(shapeIt != outShapes.end());
     const MatShape x_shape = shapeIt->second;
 
-    const int seq_length = x_shape[0];
-    const int batch_size = x_shape[1];
+    //if layout is 1, change batch and sequence dims
+    const int layout = layerParams.get<int>("layout", false);
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1757]: default value should be 0. See https://github.com/onnx/onnx/blob/906884ab594d5631cb4d31cf0d5aaeb78ed04f80/docs/Operators.md?plain=1#L19104

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 3]
```diff
@@ -1799,7 +1807,7 @@ void ONNXImporter::parseLSTM(LayerParams& layerParams, const opencv_onnx::NodePr
 
     bool need_yc = lstm_proto.output_size() > 2 && !lstm_proto.output(2).empty();
     bool need_yh = lstm_proto.output_size() > 1 && !lstm_proto.output(1).empty();
-    bool need_y = lstm_proto.output_size() > 0 && !lstm_proto.output(0).empty();
+    bool need_y  = lstm_proto.output_size() > 0 && !lstm_proto.output(0).empty();
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1810]: please revert

File: modules/dnn/test/test_onnx_importer.cpp [hunk 4]
```diff
@@ -1393,6 +1393,20 @@ TEST_P(Test_ONNX_layers, LSTM_init_h0_c0)
...
 }
+// epsilon is larger because onnx does not match with torch/opencv exactly
+TEST_P(Test_ONNX_layers, LSTM_layout_0)
+{
+    if(backend == DNN_BACKEND_CUDA)
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_CUDA);
+    testONNXModels("lstm_layout_0", npy, 0.005, 0.005, false, false, 3);
+}
+// epsilon is larger because onnx does not match with torch/opencv exactly
+TEST_P(Test_ONNX_layers, LSTM_layout_1)
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:1404]: Please reflect behavior in the test name. `LSTM_layout_0` -> `LSTM_layout_seq`, `LSTM_layout_1` -> `LSTM_layout_batch`

**Review by dkurt — CHANGES_REQUESTED** [review]
Please see comments. The main one is to introduce internal enum for different LSTM layouts  (sequence, batch) instead of a boolean flag.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:1757]: Done!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 3 above)
Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:1810]: Done! But why?

File: modules/dnn/test/test_onnx_importer.cpp (hunk 4 above)
Comment (Abdurrahheem) [reply modules/dnn/test/test_onnx_importer.cpp:1404]: Done!

File: modules/dnn/src/layers/recurrent_layers.cpp [hunk 5]
```diff
@@ -291,8 +294,13 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
         if (useTimestampDim)
         {
             CV_Assert(inp0.size() >= 2 && total(inp0, 2) == _numInp);
-            _numSamples = inp0[1];
-            outResShape.push_back(inp0[0]);
+            if (!layout){
+                _numSamples = inp0[1];
+                outResShape.push_back(inp0[0]);
+            }else{
```
[the remark this answers is not in the document]
Comment (Abdurrahheem) [reply modules/dnn/src/layers/recurrent_layers.cpp]: done!

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Just simpler for the future tracking of changed lines.

File: modules/dnn/src/layers/recurrent_layers.cpp [hunk 6]
```diff
@@ -113,12 +113,19 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
...
+    };
+
     bool useTimestampDim;
     bool produceCellOutput;
     float forgetBias, cellClip;
     bool useCellClip, usePeephole;
     bool reverse;   // If true, go in negative direction along the time axis
     bool bidirectional;  // If true, produces both forward and reversed directions along time axis
+    layout_t layout;  // If true, uses batch_size x seq_length x num_hidden for input and output, else
+                      // uses seq_length x batch_size x num_hidden
```
Comment (dkurt) [inline modules/dnn/src/layers/recurrent_layers.cpp:128]: Please change comment correspondingly to the latest changes

File: modules/dnn/src/layers/recurrent_layers.cpp [hunk 7]
```diff
@@ -349,8 +362,13 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
         if (useTimestampDim)
         {
             CV_Assert(inp0.dims >= 2 && (int)inp0.total(2) == numInp);
-            numTimeStamps = inp0.size[0];
-            numSamples = inp0.size[1];
+            if (!layout){
```
Comment (dkurt) [inline modules/dnn/src/layers/recurrent_layers.cpp:365]: `if (layout == BATCH_SEQ_HID)`

File: modules/dnn/src/layers/recurrent_layers.cpp [hunk 8]
```diff
@@ -383,6 +401,21 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
         outputs_arr.getMatVector(output);
         internals_arr.getMatVector(internals);
 
+        if (layout){
```
Comment (dkurt) [inline modules/dnn/src/layers/recurrent_layers.cpp:404]: `if (layout == BATCH_SEQ_HID)`

**Review by dkurt — APPROVED** [review]
